ops/testing.py

```diff
@@ -1759,6 +1789,104 @@ class NonAbsolutePathError(Exception):
     """
 
 
+class _MockStorage:
+    def __init__(self, location, src):
+        self._src = src
+        os.makedirs(src, exist_ok=True)
+        self._location = location
+
+    def contains(self, path):
+        return str(path).startswith(self._location)
+
+    def _srcpath(self, path):
+        suffix = path[len(self._location):]
+        if suffix[0] == '/':
+            suffix = suffix[1:]
+        srcpath = os.path.join(self._src, suffix)
+        return srcpath
+
+    def create_dir(self, path: str, make_parents: bool = False, **kwargs) -> '_Directory':
+        assert(self.contains(path))
+        srcpath = self._srcpath(path)
+
+        dirname = os.path.dirname(srcpath)
+        if not os.path.exists(dirname):
+            if not make_parents:
+                raise RuntimeError('no such directory {}'.format(dirname))
+            os.makedirs(dirname)
+        os.makedirs(srcpath)
+        return _Directory(self, path, **kwargs)
+
+    def create_file(
+            self,
+            path: str,
+            data: typing.Union[bytes, str, typing.BinaryIO, typing.TextIO],
+            encoding: typing.Optional[str] = 'utf-8',
+            make_dirs: bool = False,
+            **kwargs
+    ) -> '_File':
+        assert(self.contains(path))
```

Comment:
This is a dangerous 
https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement

Since these are disabled sometimes, I think we'd rather just have an 
```
if not self.contains(path): raise RuntimeError(...)
```
It also gives you way to give a helpful error message instead of just an assertion failure. (what is the `location` what is the `path` that was supplied, etc.)

Reply:
```
>>> def contains(foo):
...   foo.startswith('/bar')
...
>>> assert(contains('/ba'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
```
Definitely doesn't give anyone a way to debug what they did wrong.

Since many call sites are doing this, I think a helper self._ensure_path_is_contained() is reasonable.